In [1]:
import pandas as pd
from datetime import datetime


In [2]:
data = 'Customer.csv'

In [3]:
df_data = pd.read_csv('Customer.csv')

##### Mostrar las primeras 5 filas del datset

In [4]:
df_data.head()

customer_Id         DOB Gender  city_code
0       268408  02-01-1970      M        4.0
1       269696  07-01-1970      F        8.0
2       268159  08-01-1970      F        8.0
3       270181  10-01-1970      F        2.0
4       268073  11-01-1970      M        1.0

##### Mostrar las primeras filas de la columna DOB

In [5]:
print(df_data['DOB'].head())

0    02-01-1970
1    07-01-1970
2    08-01-1970
3    10-01-1970
4    11-01-1970
Name: DOB, dtype: object


##### Reemplazar todos los guiones (-) por barras (/) en la columna DBO para que las fechas coincidan con el formato que Pandas espera (%d/%m/%Y).

In [6]:
df_data['DOB'] = df_data['DOB'].str.replace('-', '/')

##### La función pd.to_datetime toma los valores de la columna 'DOB' (que pueden estar en formato de texto) y los convierte a objetos de tipo datetime. Esto permite trabajar con las fechas de manera más eficiente, realizando operaciones como cálculos de tiempo, filtrado, etc. El parámetro dayfirst=True le dice a Pandas que las fechas están en formato día/mes/año. Por ejemplo, si tienes una fecha como 02/01/1970, Pandas entenderá que el día es 02 y el mes es 01 (enero), en lugar de interpretarlo como el formato estadounidense (mes/día/año).

In [7]:
df_data['DOB'] = pd.to_datetime(df_data['DOB'], dayfirst=True)

##### Ahora que ya hemos limpiado la columna DOB (fechas de nacimiento) y la convertimos al formato datetime, el siguiente paso es continuar con la Transformación de Datos en el proceso ETL (Extracción, Transformación y Carga).

##### Revisar valores nulos o duplicados:Es importante verificar si hay datos nulos o duplicados en el dataset. Comenzamos con los valores nulos


In [8]:
df_data.isnull().sum()

customer_Id    0
DOB            0
Gender         2
city_code      2
dtype: int64

##### y luego detectamos cuantas filas duplicadas hay 

In [9]:
df_data.duplicated().sum()

np.int64(0)

##### Con np.shape() vemos el tamaño del dataset lo que nos ayudará a tomar una decisión en cuanto a los valores nulos 


In [10]:
df_data.shape

(5647, 4)

##### Un dataset con 5647 filas y 4 columnas sería considerado de tamaño moderado. No es extremadamente pequeño, pero tampoco es tan grande como para requerir técnicas avanzadas de procesamiento de grandes volúmenes de datos.

Clasificación del tamaño de datasets (en términos generales):
Pequeños: Menos de 1,000 filas.
Moderados: Entre 1,000 y 100,000 filas.
Grandes: Más de 100,000 filas.
En este caso, 5647 filas permite un análisis detallado sin dificultades de rendimiento en herramientas como Python o Power BI. Además, es un tamaño manejable para realizar ETL y exploraciones visuales de los datos sin problemas de eficiencia.

##### Para la columna 'Gender': Crearemos una nueva categoría 'Unknown' para los valores nulos:


In [11]:
df_data.fillna({'Gender': 'Unknown'}, inplace=True)

##### Para la columna 'City Code': Aquí  se puede usar un valor numérico como -1 para indicar que no hay un código de ciudad:

In [12]:
df_data.fillna({'city_code': -1}, inplace=True)

##### Verificamos nuevamente si hay valores nulos

In [13]:
df_data.isnull().sum()

customer_Id    0
DOB            0
Gender         0
city_code      0
dtype: int64

In [14]:
df_data.head()

customer_Id        DOB Gender  city_code
0       268408 1970-01-02      M        4.0
1       269696 1970-01-07      F        8.0
2       268159 1970-01-08      F        8.0
3       270181 1970-01-10      F        2.0
4       268073 1970-01-11      M        1.0

##### Guardar el DataFrame transformado en un nuevo archivo CSV

In [15]:
df_data.to_csv('Customer_transformed.csv', index=False)